C#: Make editor create NuGet fallback folder for Godot packages

neikeq opened the pull request:
Main benefits:
- Projects can be built offline. Previously you needed internet access the first time building to download the Godot packages (`Godot.NET.Sdk`).
- Changes to packages like `Godot.NET.Sdk` can be easily tested before publishing. This was already possible but required too many manual steps.
- First time builds are a bit faster, as the Sdk package doesn't need to be downloaded. In practice, the package is very small so it makes little difference.

Bumped Godot.NET.Sdk to 4.0.0-dev3 in order to enable the recent changes regarding `.mono/` -> `.godot/mono/`.

neikeq commented:
This will be backported to the 3.2 branch later, but we probably want to give it a bit more time to be tested.

akien-mga commented on modules/mono/editor/Godot.NET.Sdk/Godot.NET.Sdk/Godot.NET.Sdk_PackageVersion.txt:
```diff
@@ -0,0 +1 @@
+4.0.0-dev3
```
CI checks show a missing newline here (though GitHub doesn't).

neikeq replied:
Luckily I made the code trim that file content, so adding a new line at the end shouldn't break anything.

neikeq replied:
Yup, the line is trimmed correctly in all 3 places it's used.

akien-mga commented:
#43028 is merged, `git rebase upstream master` should remove the now merged commit from this PR.

neikeq commented:
>#43028 is merged, `git rebase upstream master` should remove the now merged commit from this PR.

I was crossing my fingers in hope GitHub would magically take care of that :smile:

neikeq commented:
Looks like I have to update my pre-commit hooks... :(

aaronfranke commented:
> `.godot/.mono/`

Note: It's actually `.godot/mono`, we don't need to hide the folder twice.

neikeq commented:
@akien-mga Are the CI static checks different from the pre-commit hooks in any way? Mine don't trigger any error. Perhaps it's my formatter version:
```
~ $ black --version
black, version 19.3b0
```

akien-mga commented:
@neikeq That one is ran manually by the CI: `misc/scripts/file_format.sh`
It's not part of pre-commit hooks because it's slow as it parses all repo files instead of the ones changed by the commit, like the pre-commit hooks do.

aaronfranke commented:
Here the failing check is "Python style checks via black (black_format.sh)"

Black is pinned to 20.8b1: https://github.com/godotengine/godot/blob/master/.github/workflows/static_checks.yml#L22

akien-mga commented:
Ah yes, black changes its mind on how to reflow long strings all the time, so it's important to use the current stable version which we're tracking (in the past we used `master` as there was no recent release and there was an important bugfix we needed, but now it's in 20.8b1).

neikeq committed:
C#: Make editor create NuGet fallback folder for Godot packages

Main benefits:
- Projects can be built offline. Previously you needed internet
  access the first time building to download the packages.
- Changes to packages like Godot.NET.Sdk can be easily tested
  before publishing. This was already possible but required
  too many manual steps.
- First time builds are a bit faster, as the Sdk package doesn't
  need to be downloaded. In practice, the package is very small
  so it makes little difference.

Bumped Godot.NET.Sdk to 4.0.0-dev3 in order to enable the
recent changes regarding '.mono/' -> '.godot/mono/'.

aaronfranke commented on modules/mono/build_scripts/godot_net_sdk_build.py:
```diff
@@ -0,0 +1,45 @@
+
+    from .solution_builder import build_solution
+
+    extra_msbuild_args = ["/p:GodotPlatform=" + env["platform"]]
```
Why not `-p:` instead of `/p:`? MSBuild seems to support it fine.

neikeq replied:
That's outside of the scope of this PR as we use `/` for MSBuild in all the other places as well. This is something good to consider changing, but we must make sure it doesn't break anything. Was the `-` always supported or was it something introduced after .NET Core?

aaronfranke commented on modules/mono/editor/GodotTools/GodotTools/GodotSharpEditor.cs:
```diff
@@ -183,6 +198,7 @@ private enum MenuOptions
         {
             CreateSln,
             AboutCSharp,
+            SetupGodotNugetFallbackFolder
```
Could add a trailing comma here.

aaronfranke commented on modules/mono/editor/GodotTools/GodotTools/Build/NuGetUtils.cs:
```diff
@@ -0,0 +1,296 @@
+  </packageSources>
+</configuration>
+";
+                    System.IO.File.WriteAllText(nuGetConfigPath, defaultConfig, Encoding.UTF8); // UTF-8 with BOM
```
Is a BOM necessary for NuGet to recognize UTF-8? It's best practice to exclude the BOM unless there is [a particular oddity](https://github.com/dotnet/sdk/issues/8929).

neikeq replied:
I checked all the `NuGet.Config` files I had and they include a BOM. I don't know if it's necessary, but considering it's the case for solution files I rather stay safe.

aaronfranke commented on modules/mono/editor/GodotTools/GodotTools.Shared/GodotTools.Shared.csproj:
```diff
@@ -0,0 +1,7 @@
+    <PropertyGroup>
+        <TargetFramework>netstandard2.0</TargetFramework>
+    </PropertyGroup>
+
```
```suggestion
```

Also, I noticed here the target is `netstandard2.0`, while Godot projects generate with `net472`, why the inconsistency? Also, there is `netstandard2.1`.

neikeq replied:
Godot projects didn't support .NET Standard at first. Today we still default to `net472` in 3.2 to avoid breaking projects, but 3.2 supports it and 4.0 defaults to `netstandard2.1`.
Regarding `GodotTools`, .NET Standard is definitely preferred, but some projects have .NET Framework dependencies like `Microsoft.Build` hence why they target .NET Framework instead (.NET Standard projects can't have .NET Framework dependencies).
The reason we target `netstandard2.0` rather than `netstandard2.1` is because the latter is only fully supported in `net480` (.NET Framework 4.8) which our projects don't target as it was too bleeding edge back then.
Also .NET Standard 2.0 is recommended over 2.1 for the best compatibility (as said before `net472` can't reference `netstandard2.1`) so `GodotTools.IdeMessaging` likely won't change as it's used outside of Godot too (VS extension).

aaronfranke commented:
I did a little testing. In the current master, building fails due to the mismatch of `.mono/` and `.godot/mono`. With this PR, pressing the build button causes the editor to crash, but the good news is that it didn't create the `.mono/` folder :P

Anyway, here's a minimal test project, including `.godot/mono/` and the logs I get when building: [Test.zip](https://github.com/godotengine/godot/files/5432548/Test.zip) For reference I'm on Ubuntu 20.04 with Mono 6.12.0.90 and .NET Core 3.1.403. I also had to copy over the `csproj` and `sln` files from a Godot 3 project and modify them, because I'm not able to open the Create Script menu without freezing my entire computer.

neikeq commented:
@aaronfranke Thanks. Custom event signals weren't doing hot-reload properly. #43088 fixes it.

aaronfranke reviewed (approved):
Tested in combination with #43088 and it works. I don't understand all this code but it looks fine.

Note: I didn't test offline builds, but at least the online builds work now unlike the current master.

akien-mga commented:
Thanks!
